Fix portrait endpoint for email address user id's

instification opened the pull request:
The previous pr #1466 resolved an issue where the endpoint was returning unsanitized user id's.

Following #1481 the new @portrait endpoint does not require sanitized user id's so when email login is enabled, portraits break.

The solution would either be to revert the change made in #1466 and return the raw user id, or to update the @portrait endpoint to handle sanitized id's.

This PR does the latter as well as:

 - Updating the endpoint for discussion/comments to use the new url
 - Updates the tests
   - Uses a Plone user instead of the zmi admin user, as this causes issues rendering the portrait
   - Adds a tests for users where login by email is enabled
   - Updates comments tests

mister-roboto commented:
@instification thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

instification commented:
@jenkins-plone-org please run jobs

tisto reviewed (approved):
LGTM. @instification thanks for the fix!

tisto commented:
@jenkins-plone-org please run jobs

jensens commented on src/plone/restapi/services/users/get.py:
```diff
@@ -222,7 +223,12 @@ def _get_user_id(self):
     def render(self):
         if len(self.params) == 1:
             # Retrieve the user portrait
-            portrait = self.portal_membership.getPersonalPortrait(self.params[0])
```
Here be dragons.

There is a long (>15y) history of patching `MembershipTool.getPersonalPortrait` in several addons dealing with it. Membrane did/does, pas.plugins.ldap as well, probably others too.

Before you merge this please check if this works with a recent version of `pas.plugins.ldap` or similar (should, but better safe than sorry).  I am also not sure if the meta_type assumption  always holds.

instification replied:
So if I understand your comment @jensens you're suggesting it would be dangerous to not use `getPersonalPortrait`?

Because other addon's may patch it to give a different result, so by not using it we could be missing out on those changes?

I can see in pas.plugins.ldap it is patching it to basically solve the issue we are seeing here: https://github.com/collective/pas.plugins.ldap/blob/e1f8da6ab300bed50c779febac9036930b2cb91f/src/pas/plugins/ldap/monkey.py#L58-L86

Unfortunately, `portal_membership.getPersonalPortrait` doesn't return a portrait if the user id is already sanitised.

Perhaps the better solution would be to 'de-sanitize' the user id before passing it to `getPersonalPortrait`?

jensens replied:
Exact, yes. And I am not sure if the meta_type is always the one you expect.

instification replied:
ok thanks for the feedback. I'll take another look at the meta_type check, I'm sure there is a better way to determine if the users portrait is unset.

instification replied:
@jensens I have refactored the code.

So now both the code that returns the portrait url and fetches the portrait will use getPersonalPortrait.

When fetching from the @portrait endpoint, it is expecting a 'safe id' so is calling `decleanId` from pas before calling `getPersonalPortrait`.

I don't really have an effective setup to test pas.ldap (and certainly not with custom portrait code), but as we are using the `getPersonalPortrait` function as intended I can't see it being an issue there due to this change.

jensens replied:
Yes, this will do it. Great work. It is always good to see others getting their fingers dirty with our membership code ;-)

instification commented:
@sneridagh I have also updated the endpoints/tests for comments as they were returning the old portrait url, which was causing it to break in volto. I can refactor the change into a new PR if you think it is better to do it that way

instification commented:
@jenkins-plone-org please run jobs

instification commented:
@jenkins-plone-org please run jobs

instification commented:
I noticed in local testing that the `isDefaultPortrait` function was always returning true, even when a users portrait wasn't set.

For some reason `getattr(portal, default_portrait, None)` always returns None, but `portal.unrestrictedTraverse(default_portrait, None)` does return the root portrait. I guess at some point in the past this was changed from being an object in zope to something else.

Additionally, even though `portal_membership.getPersonalPortrait('admin`)` returns `<FSImage /Plone/defaultUser.png>` and the traversed default portrait is also `<FSImage /Plone/defaultUser.png>` they don't compare. So I updated the function to compare on `.getPhysicalPath()` which does match for some reason.

I added a test to check for a 404 from the endpoint when the portrait isn't set.

instification commented:
@jenkins-plone-org please run jobs

davisagli commented on src/plone/restapi/tests/test_services_users.py:
```diff
@@ -1011,6 +1014,42 @@ def test_get_user_portrait_anonymous(self):
+        self.assertEqual(200, response.status_code)
+        self.assertEqual(response.headers["Content-Type"], "image/gif")
+
+    def test_get_user_defalut_portrait(self):
```
```suggestion
    def test_get_user_default_portrait(self):
```
